### ISSUE TITLE
@computed causes recalculation of all observed properties

### ISSUE THREAD

**Issue by Vladimir85** [rwjblue/ember-decorators#89]
By using Ember's computed wrapper, the function is being invoked when at least one of the observable properties has been changed. When recomputing the property, it is not forcing the recalculation of all observables.
```javascript
const { computed, get, isPresent } = Ember;

export default Ember.Object.create({
  firstProperty  : null,
  secondProperty : null,
  
  firstComputed: computed('firstProperty', 'secondComputed', 'thirdComputed', function() {
    if (isPresent(get(this, 'firstProperty'))) {
      // When this condition is verified, "thirdComputed" is not "computed"
      return get(this, 'secondComputed');
    }
    return get(this, 'thirdComputed');
  }),
  
  secondComputed: computed(function() {
    console.log('secondComputed logic executed');
    // some magic here ...
  }),
  
  thirdComputed: computed('otherProperty', function() {
    console.log('thirdComputed logic executed');
    // more magic here ...
  })
});
```
Let's assume that `firstComputed` is used at some point.
On change of the `firstProperty` the `firstComputed` will be recalculated. The `thirdComputed` will be calculated ONLY when the condition `if (isPresent(get(this, 'firstProperty')))` is `false` (exactly as it defined by the logic).

Using `ember-computed-decorators` the behavior of computed property is different.

```javascript
import C from 'ember-computed-decorators';
const { isPresent } = Ember;

export default Ember.Object.create({
  firstProperty  : null,
  secondProperty : null,
  
  @C('firstProperty', 'secondComputed', 'thirdComputed')
  firstComputed(firstProperty, secondComputed, thirdComputed) {
    // At this point all 3 properties in input are 'computed' / 'recalculated'
    if (isPresent(firstProperty)) {
      // When this condition is verified, "thirdComputed" is not "computed"
      return secondComputed;
    }
    return thirdComputed;
  },

  @C('firstProperty', 'secondComputed', 'thirdComputed')
  firstComputedNoParams() {
    // Even at this point all 3 properties in input are 'computed' / 'recalculated'
    if (isPresent(get(this, 'firstProperty'))) {
      // When this condition is verified, "thirdComputed" is not "computed"
      return get(this, 'secondComputed');
    }
    return get(this, 'thirdComputed');
  },

  @C
  secondComputed() {
    console.log('secondComputed logic executed');
    // some magic here ...
  },

  @C('otherProperty')
  thirdComputed(otherProperty) {
    console.log('thirdComputed logic executed');
    // more magic here ...
  }
});
```
On change of any observable property, all the observables will be recalculated by `firstComputed` and even by `firstComputedNoParams`.

In case the behavior is intentional, it should be documented that all the observable properties will be eagerly calculated independently of the logic from the observer property.

**Comment by pzuraq** [comment]
This appears to be because of the behavior of [ember-macro-helpers](https://github.com/kellyselden/ember-macro-helpers), which this library wraps. Ember macro helpers resolves each of the dependent keys' values and passes them into the computed function, as you're describing. It doesn't matter if the function actually uses them or not, they will always be passed in.

In order to write the same logic you have here lazily, you could do:

```
import { computed } from 'ember-decorators/object';
const { isPresent } = Ember;

export default Ember.Object.extend({
  firstProperty  : null,
  secondProperty : null,

  @computed('firstProperty', 'otherProperty')
  firstComputedNoParams() {
    // Even at this point all 3 properties in input are 'computed' / 'recalculated'
    if (isPresent(get(this, 'firstProperty'))) {
      // When this condition is verified, "thirdComputed" is not "computed"
      return get(this, 'secondComputed');
    }
    return get(this, 'thirdComputed');
  },

  @computed
  secondComputed() {
    console.log('secondComputed logic executed');
    // some magic here ...
  },

  @computed('otherProperty')
  thirdComputed(otherProperty) {
    console.log('thirdComputed logic executed');
    // more magic here ...
  }
});
```

But there are definitely cases where that wouldn't be ideal (large computeds with lots of dependent keys, for example). It looks like ember-macro-helpers provides a `lazyComputed` helper, I think it'd make sense to wrap that in this library as well.

**Comment by kellyselden** [comment]
Can confirm `lazyComputed` is what you want in this case, but more code required for the consumer and the API is not compatible. Also, before `ember-macro-helpers` was brought into this lib, it was still eagerly computing all properties.

**Comment by rwjblue** [comment]
Blocked on #141, but https://github.com/rwjblue/ember-decorators/issues/146 tracks the ultimate removal of the functionality provided by `ember-macro-helpers` to auto-get the dependencies.

Closing this in favor of #141 and #146
